src/derivkit/forecasting/forecast_core.py

```diff
@@ -0,0 +1,237 @@
+    """
+    if forecast_order not in (1, 2):
+        raise ValueError(
+            "Only Fisher (order 1) and doublet-DALI (order 2) forecasts are currently supported."
+        )
+
+    theta0_arr = np.atleast_1d(theta0)
+    cov_arr = validate_covariance_matrix(cov)
+    n_observables = cov_arr.shape[0]
+
+    # Check model output dimension
+    y0 = np.atleast_1d(function(theta0_arr))
+    if y0.shape[0] != n_observables:
+        raise ValueError(
+            f"Expected {n_observables} observables from model (from cov {cov.shape}), "
+            f"but got {y0.shape[0]} (output shape {y0.shape})."
+        )
+
+    # Compute inverse covariance matrix
+    invcov = invert_covariance(cov_arr, warn_prefix="get_forecast_tensors")
+
+    # Compute first-order derivatives
+    d1 = _get_derivatives(
+        function,
+        theta0_arr,
+        cov_arr,
+        order=1,
+        n_workers=n_workers,
+        method=method,
+        **dk_kwargs,
+    )
+
+    if forecast_order == 1:
+        return np.einsum("ai,ij,bj->ab", d1, invcov, d1)  # Fisher
+
+    # Compute second-order derivatives
+    d2 = _get_derivatives(
+        function,
+        theta0_arr,
+        cov_arr,
+        order=2,
+        n_workers=n_workers,
+        method=method,
+        **dk_kwargs,
+    )
+    # G_abc = Σ_ij d2[a,b,i] invcov[i,j] d1[c,j]
+    g_tensor = np.einsum("abi,ij,cj->abc", d2, invcov, d1)
+    # H_abcd = Σ_ij d2[a,b,i] invcov[i,j] d2[c,d,j]
+    h_tensor = np.einsum("abi,ij,cdj->abcd", d2, invcov, d2)
+    return g_tensor, h_tensor
+
+
+def _get_derivatives(
+        function: Callable[[ArrayLike], float | NDArray[np.floating]],
+        theta0: ArrayLike,
+        cov: ArrayLike,
+        *,
+         order: int,
+         method: str | None = None,
+         n_workers: int = 1,
+         **dk_kwargs: Any,
+         ) -> NDArray[np.float64]:
+    """Returns derivatives of the observables of the requested order.
+
+    Args:
+        function: The scalar or vector-valued function to
+            differentiate. It should accept a list or array of parameter
+            values as input and return either a scalar or a
+            :class:`np.ndarray` of observable values.
+        theta0: The points at which the
+            derivative is evaluated. A 1D array or list of parameter values
+            matching the expected input of the function.
+        cov: The covariance matrix of the observables. Should be a square
+            matrix with shape (n_observables, n_observables), where n_observables
+            is the number of observables returned by the function.
+        order (int): The requested order d of the derivatives:
+
+            - d = 1 returns first-order derivatives.
+            - d = 2 returns second-order derivatives.
+
+            Currently only d = 1, 2 are supported.
+
+        method: Method name or alias (e.g., "adaptive", "finite"). If None,
+            the DerivativeKit default ("adaptive") is used.
```

Comment:
```suggestion
        method: Method name or alias (e.g., ``"adaptive"``, ``"finite"``). If ``None``,
            the DerivativeKit default ("adaptive") is used.
```
I wouldn't mention what the `DerivativeKit` default is, as that might change.